lib/smack/SmackInstGenerator.cpp

```diff
@@ -145,39 +145,165 @@ void SmackInstGenerator::processInstruction(llvm::Instruction& inst) {
-    else
-      llvm_unreachable("Unexpected return type of half-precision upcast intrinsic.");
-    return Stmt::call(funcName, {rep->expr(ii->getArgOperand(0))},
-      {naming->get(*ii)});
+    // translation: $f := $fpext.bvhalf.*($rmode, $bitcast.bv16.bvhalf($i));
+    auto argT = rep->type(ii->getArgOperand(0)->getType());
+    auto retT = rep->type(ii->getFunctionType()->getReturnType());
+    assignFuncApp(SmackOptions::FloatEnabled && SmackOptions::BitPrecise,
+      "$fpext",
+      {Naming::HALF_TYPE, retT},
+      {Expr::id(Naming::RMODE_VAR),
+      Expr::fn(indexedName("$bitcast", {argT, Naming::HALF_TYPE}),
+        rep->expr(ii->getArgOperand(0)))});
   };
   static const auto f16DownCast = [&ii, this] {
-    std::string funcName;
-    auto tyID = ii->getArgOperand(0)->getType()->getTypeID();
-    if (tyID == llvm::Type::FloatTyID)
-      funcName = "__SMACK_convert_to_fp16_f32";
-    else if (tyID == llvm::Type::DoubleTyID)
-      funcName = "__SMACK_convert_to_fp16_f64";
-    else
-      llvm_unreachable("Unexpected operand type of half-precision downcast intrinsic.");
-    return Stmt::call(funcName, {rep->expr(ii->getArgOperand(0))},
-      {naming->get(*ii)});
+    // translation: assume($bitcast.bv16.bvhalf($i) == $fptrunc.bvfloat.bvhalf($rmode, $f));
+    auto argT = rep->type(ii->getArgOperand(0)->getType());
+    auto retT = rep->type(ii->getFunctionType()->getReturnType());
+    if (SmackOptions::FloatEnabled && SmackOptions::BitPrecise) {
+      emit(Stmt::assume(
+        Expr::eq(
+          Expr::fn(indexedName("$fptrunc", {argT, Naming::HALF_TYPE}),
+            Expr::id(Naming::RMODE_VAR),
+            rep->expr(ii->getArgOperand(0))),
+          Expr::fn(indexedName("$bitcast", {retT, Naming::HALF_TYPE}),
+            rep->expr(ii)))));
+    } else
+      overApproximate(ii);
+  };
+  static const auto assignBVFuncApp = [&ii, this] (std::string funcNameBase){
+    // translation: $res := $bswap.bv*($arg1);
+    auto argT = rep->type(ii->getArgOperand(0)->getType());
+    assignFuncApp(SmackOptions::BitPrecise,
+      funcNameBase,
+      {argT},
+      {rep->expr(ii->getArgOperand(0))});
+  };
+  static const auto assignUnFPFuncApp = [&ii, this]
+    // translation: $res := $<func>.bv*($arg1);
+    (std::string funcNameBase) {
+    auto argT = rep->type(ii->getArgOperand(0)->getType());
+    assignFuncApp(SmackOptions::FloatEnabled,
+      funcNameBase,
+      {argT},
+      {rep->expr(ii->getArgOperand(0))});
+  };
+  static const auto assignBinFPFuncApp = [&ii, this]
+    // translation: $res := $<func>.bv*($arg1, $arg2);
+    (std::string funcNameBase) {
+    auto argT = rep->type(ii->getArgOperand(0)->getType());
+    assignFuncApp(SmackOptions::FloatEnabled,
+      funcNameBase,
+      {argT},
+      {rep->expr(ii->getArgOperand(0)), rep->expr(ii->getArgOperand(1))});
+  };
+  static const auto assignRoundFPFuncApp = [&ii, this]
+    // translation: $res := $round.bv*(rmode, $arg1);
```

Comment:
It is weird that comment comes before the signature is finished. Fix this formatting.